### PR TITLE
test(core): enforce documented architecture rules

### PR DESCRIPTION
**Pull request by bojtospeter**
## Summary

- turn every documented architecture rule into an executable guard with an explicit `@Test` binding
- add negative fixtures that prove visibility, package, logging, and testing rules reject violations
- cover current core and distro classes, including parameterized and nested test categories
- make the documented command build current production artifacts and fail when zero tests execute

## Changes

- adds PO-2 configuration placement and TR-4 test-base enforcement
- reconciles documented rule sections and summary tables with executable test methods
- configures Surefire to reject empty and unmatched test selections
- updates architecture commands in contributor guidance and the PR template

## Test plan

- [x] `mvn clean install -DskipTests`
- [x] `mvn install -DskipTests -pl data-migrator/distro -am && mvn test -Pintegration -pl data-migrator/qa/integration-tests -Dtest=ArchitectureTest` (36 tests)
- [x] nonexistent `-Dtest` selector exits non-zero

## Baseline

Built from commit: `a7da8db88fb5b1c9fad5e778c5c37403cc6a16e9`

related to #1880

### CODE REVIEW

File: data-migrator/qa/integration-tests/src/test/java/io/camunda/migration/data/qa/ArchitectureTest.java [hunk 1]
```diff
@@ -33,11 +71,30 @@
  */
 class ArchitectureTest {
 
+  protected static final ImportOption EXCLUDE_ARCHITECTURE_FIXTURES =
+      location -> !location.contains("/architecturefixture/");
```
[the remark this answers is not in the document]
Comment (bojtospeter) [reply data-migrator/qa/integration-tests/src/test/java/io/camunda/migration/data/qa/ArchitectureTest.java]: Good catch — fixed in 39782572. Fixture path matching now accepts both Unix and Windows separators, with an explicit cross-platform regression test.

File: data-migrator/qa/integration-tests/src/test/java/io/camunda/migration/data/qa/ArchitectureTest.java [hunk 2]
```diff
@@ -179,29 +300,199 @@ void shouldHaveProtectedAutowiredFields() {
...
+  }
+
+  protected static void checkShouldNotUseSystemOutOrPrintStackTrace(JavaClasses classesToCheck) {
     noClasses()
         .that().resideInAPackage("io.camunda.migration.data..")
-        .should().callMethod(System.class, "out")
+        .and().resideOutsideOfPackage("..app..")
+        .should().accessField(System.class, "out")
         .orShould().callMethod("java.lang.Throwable", "printStackTrace")
         .allowEmptyShould(true)
```
[the remark this answers is not in the document]
Comment (bojtospeter) [reply data-migrator/qa/integration-tests/src/test/java/io/camunda/migration/data/qa/ArchitectureTest.java:320]: Good catch — fixed in f1bc1d1d. LR-3 now excludes the QA package, so its implementation matches the documented production-code scope.

File: data-migrator/qa/integration-tests/src/test/java/io/camunda/migration/data/qa/ArchitectureTest.java [hunk 3]
```diff
@@ -179,29 +300,199 @@ void shouldHaveProtectedAutowiredFields() {
...
+  }
+
+  protected static void checkShouldHaveTestSuffixForTestClasses(JavaClasses classesToCheck) {
     classes()
         .that().resideInAPackage("..qa..")
-        .and().haveModifier(JavaModifier.PUBLIC)
         .and().areNotInterfaces()
         .and().areNotEnums()
         .and().areNotAnnotations()
+        .and().doNotHaveModifier(JavaModifier.ABSTRACT)
```
[the remark this answers is not in the document]
Comment (bojtospeter) [reply data-migrator/qa/integration-tests/src/test/java/io/camunda/migration/data/qa/ArchitectureTest.java:354]: Thanks for checking this edge case. No change is needed here: ArchUnit `JavaClass.getMethods()` inspects methods declared by the class, while inherited methods are exposed through `getAllMethods()`. The named `MultiTenancyTest` nested containers are included in the current production scan and the 37-test architecture suite passes, confirming inherited tests do not trigger TR-5.

File: data-migrator/qa/integration-tests/src/test/java/io/camunda/migration/data/qa/ArchitectureTest.java [hunk 4]
```diff
@@ -318,15 +612,21 @@ public void check(JavaClass javaClass, ConditionEvents events) {
...
+            // Allow ProcessEngineConfigurationImpl from test lifecycle and test methods
             if (access.getOrigin() instanceof com.tngtech.archunit.core.domain.JavaMethod originMethod) {
-              boolean isTestMethod = originMethod.getAnnotations().stream()
+              boolean isTestInfrastructureMethod = isJUnitTestMethod(originMethod) ||
+                  originMethod.getAnnotations().stream()
                   .anyMatch(annotation ->
                       annotation.getRawType().getName().equals("org.junit.jupiter.api.BeforeEach") ||
-                      annotation.getRawType().getName().equals("org.junit.jupiter.api.AfterEach") ||
-                      annotation.getRawType().getName().equals("org.junit.jupiter.api.Test"));
+                      annotation.getRawType().getName().equals("org.junit.jupiter.api.AfterEach"));
```
[the remark this answers is not in the document]
Comment (bojtospeter) [reply data-migrator/qa/integration-tests/src/test/java/io/camunda/migration/data/qa/ArchitectureTest.java:636]: Good catch — fixed in c631f27d. The TR-2 exception now includes `@BeforeAll` and `@AfterAll`, with a regression fixture proving engine-configuration access is accepted from both hooks.

File: docs/TESTING_GUIDELINES.md [hunk 5]
```diff
@@ -87,6 +87,8 @@ void shouldSaveSkipReasonToDatabase() {
 
 **Exceptions:**
 - `org.camunda.bpm.engine.impl.util.ClockUtil` - allowed for time manipulation
+- `org.camunda.bpm.engine.impl.cfg.ProcessEngineConfigurationImpl` - allowed from test lifecycle
+  and test methods when engine configuration is part of the scenario
```
[the remark this answers is not in the document]
Comment (bojtospeter) [reply docs/TESTING_GUIDELINES.md:94]: Good catch — fixed in 2ce5ee65. The TR-2 rule and rationale now explicitly cover `ClockUtil`, scenario-specific `ProcessEngineConfigurationImpl` access, and `@WhiteBox` tests.

File: data-migrator/qa/integration-tests/src/test/java/io/camunda/migration/data/qa/ArchitectureTest.java [hunk 6]
```diff
@@ -49,23 +113,49 @@ void shouldNotAccessImplClasses() {
...
+        "C:\\workspace\\io\\camunda\\migration\\data\\architecturefixture\\Invalid.class").matches())
+        .isTrue();
+  }
+
+  @Test
+  void shouldAllowEngineConfigurationAccessFromAllLifecycleMethods() {
+    JavaClasses fixtureClasses = new ClassFileImporter().importClasses(EngineImplLifecycleAccessTest.class);
+
+    checkShouldNotAccessCamundaBpmEngineImplClasses(fixtureClasses);
   }
```
[the remark this answers is not in the document]
Comment (bojtospeter) [reply data-migrator/qa/integration-tests/src/test/java/io/camunda/migration/data/qa/ArchitectureTest.java:151]: Good catch — fixed in 2ce5ee65. The lifecycle fixture now exercises `@BeforeAll`, `@AfterAll`, `@BeforeEach`, and `@AfterEach`, matching the test name and the complete implemented lifecycle surface.
